src/Features/Core/Portable/EmbeddedLanguages/StackFrame/StackFrameParser.cs

```diff
@@ -0,0 +1,455 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.EmbeddedLanguages.Common;
+using Microsoft.CodeAnalysis.EmbeddedLanguages.VirtualChars;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.EmbeddedLanguages.StackFrame
+{
+    using StackFrameNodeOrToken = EmbeddedSyntaxNodeOrToken<StackFrameKind, StackFrameNode>;
+    using StackFrameToken = EmbeddedSyntaxToken<StackFrameKind>;
+    using StackFrameTrivia = EmbeddedSyntaxTrivia<StackFrameKind>;
+
+    /// <summary>
+    /// Attempts to parse a stack frame line from given input. StackFrame is generally
+    /// defined as a string line in a StackTrace. See https://docs.microsoft.com/en-us/dotnet/api/system.environment.stacktrace for 
+    /// more documentation on dotnet stack traces. 
+    /// </summary>
+    internal partial struct StackFrameParser
+    {
+        private StackFrameParser(VirtualCharSequence text)
+        {
+            _lexer = new(text);
+        }
+
+        private StackFrameLexer _lexer;
+        private StackFrameToken CurrentCharAsToken() => _lexer.CurrentCharAsToken();
+
+        /// <summary>
+        /// Given an input text, and set of options, parses out a fully representative syntax tree 
+        /// and list of diagnostics.  Parsing should always succeed, except in the case of the stack 
+        /// overflowing.
+        /// </summary>
+        public static StackFrameTree? TryParse(VirtualCharSequence text)
+        {
+            if (text.IsDefault)
+            {
+                return null;
+            }
+
+            try
+            {
+                return new StackFrameParser(text).TryParseTree();
+            }
+            catch (InsufficientExecutionStackException)
+            {
+                return null;
+            }
+        }
+
+        /// <summary>
+        /// Constructs a <see cref="VirtualCharSequence"/> and calls <see cref="TryParse(VirtualCharSequence)"/>
+        /// </summary>
+        public static StackFrameTree? TryParse(string text)
+            => TryParse(VirtualCharSequence.Create(0, text));
+
+        /// <summary>
+        /// Attempts to parse the full tree. Returns null on malformed data
+        /// </summary>
+        private StackFrameTree? TryParseTree()
+        {
+            var (_, methodDeclaration) = TryParseMethodDeclaration();
```

Comment:
if this is the only place you call TryParseMethodDeclaration, and you are always skipping the success value, then don't bother with a ParseResult here.

Reply:
I put like this because null is invalid here, but I think checking success as well will make this code more maintainable. Changing